Add weather station to opensesame rust code and doc

fel115 opened the pull request:
Added rust code for connecting to the weather station and some documentation of the process.

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,21 @@
+# Weather Station
+This documentation outlines the chosen steps for establishing communication with the CLIMA SENSOR US 4.920x.x0.xxx weather station. The hardware utilized for this communication includes an [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware) board with [LIME2-SHIELD](https://www.olimex.com/Products/OLinuXino/A20/LIME2-SHIELD/open-source-har ...
```
Please use one line per sentence, it is easier to review. Always write what we use (RTU vs. ASCII).

Suggestion only fixes one line per sentence:

```suggestion

This documentation outlines the chosen steps for establishing communication with the CLIMA SENSOR US 4.920x.x0.xxx weather station.
The hardware utilized for this communication includes an [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware) board with [LIME2-SHIELD](https://www.olimex.com/Products/OLinuXino/A20/LIME2-SHIELD/open-source-hardware) and a [MOD-RS485](https://www.olimex.com/Products/Modules/Interface/MOD-RS485/open-source-hardware) module.
Two command interpreters, namely THIES-ASCII and MODBUS RTU, can be chosen for communication, with THIES-ASCII being the default option.
```

fel115 replied:
with next commit resolved

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,21 @@
+## Initial Communication Steps
+
+### Pin Identification
+To identify the correct pins, refer to the datasheet. In our case, we used the `UEXT2` pins on the LIME2-Shield. We discovered that the `TX` and `RX` pins of `UEXT2` are connected to the `uart7` of the A20-SoC. In the next step, we searched the device tree for `uart7 (serial@1c29c00)` to determine its address. By checking the system logs using `dmesg | grep '1c29c00'`, we found that `uart7` is ma ...
```
This should write what users need to do and they shouldn't need the datasheet.

```suggestion

We used the `UEXT2` pins on the LIME2-Shield.
The `TX` and `RX` pins of `UEXT2` are connected to the `uart7` of the A20-SoC.
In the next step, we searched the device tree for `uart7 (serial@1c29c00)` to determine its address.
By checking the system logs using `dmesg | grep '1c29c00'`, we found that `uart7` is mapped to `/dev/ttyS5`.
```

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,21 @@
+
+Currently, we are able to observe output on the RS485 bus, but we have not received a response. This could be due to the weather station being set to full-duplex by default, whereas we can only use half-duplex.
+
+To compile the Rust code, you need to install the `libudev` library. During cross-compilation, we encountered issues with this library, so we opted to compile it directly on the [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware).
```
How was it compiled?

fel115 replied:
Do you mean `libudev`? I downloaded `libudev` from https://packages.debian.org/bullseye/libudev-dev and installed it with `deb -i ...`

markus2330 replied:
So I misunderstood it, I understood the sentence that you compiled `libudev` "directly on theA20-OLinuXino-LIME2". Please rewrite what you did step by step. Btw. why not `apt install libudev-dev`?

markus2330 commented on src/weather_station/main.rs:
```diff
@@ -0,0 +1,46 @@
+    let mut gpio_scl = gpio::sysfs::SysFsGpioOutput::open(273).unwrap();
+    let mut gpio_ss = gpio::sysfs::SysFsGpioOutput::open(272).unwrap();
+
+    // Öffne die serielle Schnittstelle
```
English comments please

markus2330 commented on src/weather_station/main.rs:
```diff
@@ -0,0 +1,46 @@
+    let device = &args[1];
+    let baudrate: u32 = args[2].parse().unwrap();
+
+    let mut gpio_scl = gpio::sysfs::SysFsGpioOutput::open(273).unwrap();
```
Don't use magic constants, introduce a constant saying what 273 is.

markus2330 reviewed (changes requested):
Thanks for the PR!

markus2330 commented on src/weather_station/main.rs:
```diff
@@ -0,0 +1,46 @@
+    // Sende Daten über die serielle Schnittstelle
+    gpio_scl.set_value(true).expect("cant set gpio_scl to true");
+    gpio_ss.set_value(true).expect("cant set gpio_ss to true");
+
```
Are you sure that here is no delay required? Do the set_value work immediately? Will the Mod-RS485 switch its mode without delay?

fel115 replied:
This can be handled with this function `modbus_rtu_set_custom_rts` of libmodbus.

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -0,0 +1,19 @@
+# libmodbus Trouble Protocol
```
Please write separate documents for the goodwe-inverter and the thiesclima-weatherstation.

In there write about expected electrical wiring, [RS485](https://www.olimex.com/Products/Modules/Interface/MOD-RS485/) jumpers settings (there are all default still, please check.)

You can keep this file also for general troubleshooting (lib)modbus.

markus2330 replied:
I just saw that for thiesclima-weatherstation there is already a file :+1:.

Please add the requested content there.

fel115 replied:
I created a separate doc file for goodwe-inverter, but for now, everything is quite the same. Only the `SLAVE_ID`is different. And in the thiesclima doc, I mainly wrote about the steps I took, to get where I am now. So maybe I create a new doc file with the trial and error process and then two files, one for the weather station and one for the inverter. And in those two files, I put information about which methods are for reading and writing which register.  Because I think the thiesclima doc has too much unnecessary information in it.

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -0,0 +1,19 @@
+# libmodbus Trouble Protocol
+
+## config modbus-rtu connection 
+1. `ctx = modbus_new_rtu("/dev/ttyS5", 9600, 'N', 8, 1)` - set ttyS5 for serial communication; Baudrate to 9600; No paritybit; 8 databit; 1 stopbit
```
This should be rather directly in code. Please write where in the instructions/manuals you found these settings or on which places there is still speculation and thus might be wrong.

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -0,0 +1,19 @@
+## config modbus-rtu connection 
+1. `ctx = modbus_new_rtu("/dev/ttyS5", 9600, 'N', 8, 1)` - set ttyS5 for serial communication; Baudrate to 9600; No paritybit; 8 databit; 1 stopbit
+2. `modbus_rtu_set_serial_mode(ctx, MODBUS_RTU_RS485)`- set serial mode to RS485, other option would be RS323
+3. `modbus_rtu_set_rts(ctx, MODBUS_RTU_RTS_UP)` - set the RTS-PINS (9&10) to high while sending, and while receiving low
```
Isn't this done by custom code?

fel115 replied:
no, the header of the custom function looks like `void set_rts_custom(modbus_t *ctx, int on)`, and the on parameter indicates what should be done. For example, if we set `MODBUS_RTU_RTS_UP` then `on=1` before sending and after sending the same function is called with `on=0` so we can differ between before and after sending. With `MODBUS_RTU_RTS_UP` it would be the other way

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -0,0 +1,19 @@
+1. `ctx = modbus_new_rtu("/dev/ttyS5", 9600, 'N', 8, 1)` - set ttyS5 for serial communication; Baudrate to 9600; No paritybit; 8 databit; 1 stopbit
+2. `modbus_rtu_set_serial_mode(ctx, MODBUS_RTU_RS485)`- set serial mode to RS485, other option would be RS323
+3. `modbus_rtu_set_rts(ctx, MODBUS_RTU_RTS_UP)` - set the RTS-PINS (9&10) to high while sending, and while receiving low
+4. `modbus_set_slave(ctx, REMOTE_ID)` - set the slave-id
```
Where did you read about the slave-id? Could it be that it was changed?

fel115 replied:
I read about it in the thiesclima documentation, in the section 'Befehlsliste/Befehl ID', there is the default `ID` set to `0`. But now I know it is configured with the `SLAVE_ID=1`

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -0,0 +1,19 @@
+2. `modbus_rtu_set_serial_mode(ctx, MODBUS_RTU_RS485)`- set serial mode to RS485, other option would be RS323
+3. `modbus_rtu_set_rts(ctx, MODBUS_RTU_RTS_UP)` - set the RTS-PINS (9&10) to high while sending, and while receiving low
+4. `modbus_set_slave(ctx, REMOTE_ID)` - set the slave-id
+5. `modbus_rtu_set_custom_rts(ctx, &set_rts_custom)` - set a custom function which is called befor and after sending data
```
```suggestion
5. `modbus_rtu_set_custom_rts(ctx, &set_rts_custom)` - set a custom function which is called before and after sending data
```

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -0,0 +1,19 @@
+3. `modbus_rtu_set_rts(ctx, MODBUS_RTU_RTS_UP)` - set the RTS-PINS (9&10) to high while sending, and while receiving low
+4. `modbus_set_slave(ctx, REMOTE_ID)` - set the slave-id
+5. `modbus_rtu_set_custom_rts(ctx, &set_rts_custom)` - set a custom function which is called befor and after sending data
+6. `modbus_connect(ctx)` - starts the connection with the serial connection
```
too code like, I don't think we need this here in the documentation. Documentation should be what is relevant for user: what to check if there is no connection like we have now.

```suggestion
6. `modbus_connect(ctx)` - starts the serial connection
```

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -0,0 +1,19 @@
+## weatherstation
+### Date 05.07.2023
+- `Connection timed out` with the above config and the `SLAVE_ID=0`
+- `Illegal data value` with the above config and the `SLAVE_ID=1`
```
Wow great, this sounds like something! Didn't know that we already got another error.

Also here: more details what you have executed and the result. You can simply say you executed sudo ./connection, then it is obvious that you mean the source with the same commit as the documentation is.

Do we already have some conclusion about what this error means? To document this might help future people with the same problem.

markus2330 commented on doc/libmodbus-troubles.md:
```diff
@@ -0,0 +1,19 @@
+
+## weatherstation
+### Date 05.07.2023
+- `Connection timed out` with the above config and the `SLAVE_ID=0`
```
please write which program produced which exact behavior (afaik read_registered timed out, and never write_registers. This document currently implies the opposite.)

fel115 replied:
Both methods can have a time-out. Because if you write a register the slave also answers, as acknowledgment.

markus2330 replied:
Didn't the weatherstation have successful writes (changing full-duplex mode)? The protocol should answer such questions.

fel115 replied:
Writing wasn't successful, because I used the wrong `SLAVE_ID`. But with the right `SLAVE_ID` writing should work.

markus2330 replied:
Great, please create a new commit with the minimal working examples.

markus2330 commented on src/pv_anlage/connection.c:
Please rename to "src/goodwe-inverter/connection.c". Please also add Makefile/compilation script.

markus2330 commented on src/pv_anlage/connection.c:
```diff
@@ -0,0 +1,177 @@
+
+void config_gpio_pins(){
+     // Export the desired pin by writing to /sys/class/gpio/export
+    int fd = open("/sys/class/gpio/export", O_WRONLY);
+    if (fd == -1) {
+        perror("Unable to open /sys/class/gpio/export");
+        exit(1);
+    }
+
+    if (write(fd, "272", 3) != 3) {
+        perror("Error writing to /sys/class/gpio/export");
+        exit(1);
+    }
+
+    close(fd);
```
put this to a function export_gpio or similar, this would make the config_gpio_pins function much easier to read.

fel115 replied:
The code isn't final. I will clean everything up if I get an answer from the inverter.

markus2330 replied:
My experience is that it is much more efficient debugging hardware problems if you have a code that you are confident in.

markus2330 commented on src/pv_anlage/connection.c:
```diff
@@ -0,0 +1,177 @@
+#include <stdio.h>
```
Please add some (short) general info about the file: what it does etc.

markus2330 commented on src/pv_anlage/connection.c:
```diff
@@ -0,0 +1,177 @@
+        return -1;
+    }
+
+    // Read 2 registers from address 0 of server ID 10.
```
is this up-to-date?

markus2330 commented on src/pv_anlage/connection.c:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    // Read 2 registers from address 0 of server ID 10.
+    if (modbus_read_registers(ctx, 0x0200, 0x0008, tab_reg) == -1){
```
Please avoid magic constants, create a const variable for 0x0200 and 0x0008 which name describe what registers/sizes it contains.  (like REMOTE_ID)

markus2330 reviewed (changes requested):
Thx, `Illegal data value` already looks promising, hopefully this wasn't only wrong data passed in the program but something we actually received? Looking forward to further experiments.

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+# Weather Station
+This documentation outlines the chosen steps for establishing communication with the CLIMA SENSOR US 4.920x.x0.xxx weather station. 
+The hardware utilized for this communication includes an [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware) board with [LIME2-SHIELD](https://www.olimex.com/Products/OLinuXino/A20/LIME2-SHIELD/open-source-hardware) and a [MOD-RS485](https://www.olimex.com/Products/Modules/Interface/MOD-RS485/open-source-hardware) module. 
```
```suggestion
The hardware utilized for this communication includes an [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware) board with [LIME2-SHIELD](https://www.olimex.com/Products/OLinuXino/A20/LIME2-SHIELD/open-source-hardware) and a [MOD-RS485](https://www.olimex.com/Products/Modules/Interface/MOD-RS485/open-source-hardware) module.
```

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+This documentation outlines the chosen steps for establishing communication with the CLIMA SENSOR US 4.920x.x0.xxx weather station. 
+The hardware utilized for this communication includes an [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware) board with [LIME2-SHIELD](https://www.olimex.com/Products/OLinuXino/A20/LIME2-SHIELD/open-source-hardware) and a [MOD-RS485](https://www.olimex.com/Products/Modules/Interface/MOD-RS485/open-source-hardware) module. 
+Two command interpreters, namely THIES-ASCII and MODBUS RTU, can be chosen for communication, with THIES-ASCII being the default option. 
+We will be utilizing the THIES-ASCII interpreter for this project.
```
Is this still up-to-date?

```suggestion
We will be utilizing the THIES-ASCII command interpreter.
```

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+## Initial Communication Steps
+
+### Pin Identification
+We used the `UEXT2` pins on the LIME2-Shield.
```
Use present tense.

```suggestion
We connect MOD-RS485 via [UEXT](https://www.olimex.com/Products/Modules/) on UEXT2 of the LIME2- Shield (UEXT1 can be used for MOD-ENV).
```

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+### Using mbpoll
+Initially, we attempted to communicate with the [mbpoll](https://github.com/epsilonrt/mbpoll) tool. 
+However, we encountered an issue where the weather station had the default address `0x00`, and mbpoll was unable to send Modbus RTU messages to this address. 
+To resolve this, we had to make some [minor modifications](https://github.com/epsilonrt/mbpoll/issues/39) to the source code of mbpoll and recompile it, enabling communication with address `0x00`.
```
This sounds like it was tried and worked? Please clarify in the text.

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+However, we encountered an issue where the weather station had the default address `0x00`, and mbpoll was unable to send Modbus RTU messages to this address. 
+To resolve this, we had to make some [minor modifications](https://github.com/epsilonrt/mbpoll/issues/39) to the source code of mbpoll and recompile it, enabling communication with address `0x00`.
+
+### Developing Custom Rust Code
+
+In our Rust code, we utilize the `serialport` and `gpio` modules. 
+The `serialport` module allows us to send serial packages to the MOD-RS485 board, while the `gpio` module enables read and write operations by configuring the SCL/SCK (GPIO273) and #SS/SDA (GPIO272) pins. 
+When both pins are set to `0`, the MOD-RS485 is ready to receive packages, and when both are set to `1`, the MOD-RS485 is ready to send packages.
+
+Currently, we are able to observe output on the RS485 bus, but we have not received a response. This could be due to the weather station being set to full-duplex by default, whereas we can only use half-duplex.
+
+To compile the Rust code, you need to install the `libudev` library. 
+During cross-compilation, we encountered issues, so we opted to compile it directly on the [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware).
```
I think this is irrelevant now? (libmodbus as the way to go?)

```suggestion
```

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+During cross-compilation, we encountered issues, so we opted to compile it directly on the [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware).
+
+#### ASCII
+The initial attempt was to send ASCII messages over the serial connection, as implemented in the `src/weather_station/connection_ascii.rs` file. In the `connection_ascii.rs` file, we attempted to send data to configure the weather station for half-duplex usage. Although we were able to cache the output on the RS-485 bus, we did not receive a response from the weather station. This could be due to ...
```
Please one line per sentence. Is this still up to date?

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+The initial attempt was to send ASCII messages over the serial connection, as implemented in the `src/weather_station/connection_ascii.rs` file. In the `connection_ascii.rs` file, we attempted to send data to configure the weather station for half-duplex usage. Although we were able to cache the output on the RS-485 bus, we did not receive a response from the weather station. This could be due to ...
+
+#### RTU-Modbus
+The second attempt involved sending data using the libmodbus library, as implemented in the `src/weather_station/connection_modbus.rs` file. For this Modbus connection, we utilized [libmodbus-rs](https://github.com/zzeroo/libmodbus-rs), which provides an implementation of most of the methods from libmodbus. However, we encountered an issue while using [libmodbus-rs](https://github.com/zzeroo/libm ...
```
The C program is hopefully only a temporary solution until https://github.com/zzeroo/libmodbus-rs/issues/18 is fixed. Once we have a working C program setting the two pins, please add this to ttps://github.com/zzeroo/libmodbus-rs/issues/18 so that zzeroo sees what we want to achieve. (Maybe there is a better way than `rtu_set_custom_rts`).

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+Another approach we considered was utilizing minicom for serial communication since it provides additional options for RS-485. However, we encountered the limitation of being unable to configure the SCL/SCK and #SS/SDA pins in minicom.
+
+## Troubles
+### MOD-RS485
+When using the [MOD-RS485](https://www.olimex.com/Products/Modules/Interface/MOD-RS485/open-source-hardware) in half-duplex mode, you have to switch SCK and #SS of the UEXT connection. 
```
Imho this would be simply a bug on our side if this isn't done.

```suggestion
```

markus2330 commented on doc/Wetterstation.md:
```diff
@@ -0,0 +1,46 @@
+### MOD-RS485
+When using the [MOD-RS485](https://www.olimex.com/Products/Modules/Interface/MOD-RS485/open-source-hardware) in half-duplex mode, you have to switch SCK and #SS of the UEXT connection. 
+
+### libmodbus-rs
+## read_register timeout 
+Because switching of SCK and #SS isn't enabled, so we need to use `rtu_set_custom_rts`, which is not implemented in [libmodbus-rs](https://github.com/zzeroo/libmodbus-rs) so wie need to programm this connection in `C`.
```
I don't think this will help someone who has timeouts.

markus2330 commented on src/weather_station/connection.c:
```diff
@@ -0,0 +1,198 @@
+#include <stdio.h>
```
please write in all files what they are supposed to do. The name "connection" is also very general.

fel115 replied:
I didn't perform that task because the C files were intended for testing the connection to the slaves. Therefore, I don't see any reason to do that right now.

markus2330 replied:
Something like this is enough:

```suggestion
/// Test connection to Thies Weather station by reading a register

#include <stdio.h>
```

markus2330 commented on src/weather_station/connection_ascii.rs:
```diff
@@ -0,0 +1,53 @@
+    let change_to_admin = b"00KY4711";
+    port.write_all(change_to_admin)?;
+
+    //sleep dazwischen
```
?

markus2330 commented on src/weather_station/connection_ascii.rs:
```diff
@@ -0,0 +1,53 @@
+use serialport::{DataBits, Parity, StopBits, Error};
```
And most important: wrote if these programs ever worked.

markus2330 reviewed (changes requested):
Review of the rest (not only latest commit).

markus2330 commented on Cargo.toml:
```diff
@@ -65,3 +65,11 @@ gettext-rs = "0.7.0"
 sunrise = "1.0.0"
+
+serialport = "4.2.1"
+gpio = "0.4.1"
```
we already use gpio-cdev

fel115 replied:
Yes, I am going to remove both, because the serialport is only needed for ASCII-mode and we use RTU MODBUS

markus2330 replied:
Please add the `libmodbus-rs` in another PR, otherwise this will never be finished. This is only to merge docu + "hello world" C programs.

fel115 replied:
Okey

markus2330 commented on doc/goodwe-inverter.md:
```diff
@@ -0,0 +1,9 @@
+# Goodwe inverter
+his documentation outlines the chosen steps for establishing communication with a goodwe-inverter. 
```
```suggestion

This documentation outlines the chosen steps for establishing communication with a goodwe-inverter. 
```

markus2330 commented on doc/goodwe-inverter.md:
```diff
@@ -0,0 +1,9 @@
+The hardware utilized for this communication includes an [A20-OLinuXino-LIME2](https://www.olimex.com/Products/OLinuXino/A20/A20-OLinuXino-LIME2/open-source-hardware) board with [LIME2-SHIELD](https://www.olimex.com/Products/OLinuXino/A20/LIME2-SHIELD/open-source-hardware) and a [MOD-RS485](https://www.olimex.com/Products/Modules/Interface/MOD-RS485/open-source-hardware) module. 
+
+## important information
+The `SLAVE_ID` is default `247`.
```
```suggestion

The `SLAVE_ID` by default is `247`.
```

markus2330 commented on src/goodwe-inverter/connection.c:
```diff
@@ -0,0 +1,197 @@
+/*This connection-progam tries to start a connection with the goodwe-inverter
+
+importent: the SLAVE_ID for the inverter is 247; baudate: 9600; Bits 8N1; one register has 16bit or 4byte
+ 
```
as this is hard-coded below anyway...

```suggestion
 
```

markus2330 commented on src/goodwe-inverter/connection.c:
```diff
@@ -0,0 +1,197 @@
+/*This connection-progam tries to start a connection with the goodwe-inverter
```
please describe what exactly it does (reading register XXX...)

markus2330 commented on src/weather_station/connection.c:
```diff
@@ -0,0 +1,197 @@
+-> example registers are  Befehl AV (40015); Befehl (40031); from Page 40 of the doc which is linked in the thieclima-weatherstation.md file
+
+execution: ./connection <register> <size of reg>
+[INFO]: only Holding Register possible */
```
What does that mean?

fel115 replied:
I'm uncertain about what you're not comprehending.

markus2330 replied:
What "Holding Register" means. CU in a minute

markus2330 reviewed (changes requested):
A few last changes, let us get this done so that we can merge.

fel115 commented:
Now I have added all changes, and it is ready to merge. I removed the changes in the `Cargo.toml` file, because  implemented the hello_world in C and not in Rust

markus2330 commented:
Great job that we know that both devices worked. Looking forward to see a constant stream of data soon :heart:
